### ISSUE TITLE
Subtitles timing is off

### ISSUE THREAD

**Issue by svalavuo** [aajanki/yle-dl#360]
Is it just me and my setup or does the latest yle-dl (20240429) make subtitles appear earlier than they should?
Tried with Vera Stanhope (https://areena.yle.fi/1-2503613) all episodes. And same with Roy Grace (https://areena.yle.fi/1-50791118) all episodes.

- Samuli

**Comment by svalavuo** [comment]
I forgot to tell which players I used, but VLC and Jellyfin (both client application and web client) behaved the same. Quite odd, that previous showings of the same episodes worked fine.

**Comment by IlmariKu** [comment]
@svalavuo I had the exact same issue with a series. Did you download the Vera Stanhope as series? I noticed with independent files they seemed to be on time, but with series, lagging for few seconds. I'm also using Jellyfin.

**Comment by svalavuo** [comment]
@IlmariKu Ah, ok. I downloaded both those series as series. I have to try to download as individual episodes. Thanks for the tip!

**Comment by svalavuo** [comment]
Ok. That didn't help in my case. :-( When I downloaded those as series, subtitles were a bit early. Now that I downloaded those as individual episodes, subtitles are a bit late.

**Comment by IlmariKu** [comment]
I wonder @svalavuo, if you execute the command with `--showmetadata`-flag, it should give you the direct url of subs as a link to Areena CDN. If you download those subs, are they on-time with the video?

**Comment by svalavuo** [comment]
There is something funny going with Areena. :-) 
When I execute command yle-dl --showmetadata https://areena.yle.fi/1-67480272 (Roy Grace S04E01), there is this: "subtitles": []
And at the very beginning there is an error row: [hls @ 0xe9ac40] Can't support the subtitle(uri: ../hdntl=exp=...

**Comment by aajanki** [comment]
Not all videos have external subtitle files. It seems that especially recent Areena programs do not have external subtitles anymore.

Embedded subtitles do not show up in the --showmetadata output, only external subtitles when they are available.

It is safe to ignore the "Can't support the subtitle" warning. Ffmpeg outputs it on all Areena streams but still downloads the (embedded) subtitles.

**Comment by svalavuo** [comment]
There is no fix for this?
I was thinking that this might work if I rip the subtitles from mkv and try to fix the timing of this external subtitle and leave it external  (and disable the embedded track) or embed it back to mkv, That would take a huge amount of time with only one season of one series. :-(

**Comment by svalavuo** [comment]
Now, what I have tried with, they are all quite close two seconds early. So I delayed subtitles that 2 seconds with mkvtoolnix. Seems to work.

**Comment by IlmariKu** [comment]
Which hardware / OS are you running? I'm using Raspberry pi with Ubuntu. I wonder if the 2 seconds is caused by something related to yle-dl or maybe some slowness in hardware.

**Comment by svalavuo** [comment]
Apparently my answer wasn't saved earlier. :-(
I use mainly Asustor AS4604T NAS and Odroid M1. Both do the same: video saves nicely and subtitles timing is about exactly two seconds off. I use Subtitle Edit to determine the offset and then MKVToolNix to insert delay to subtitle track.
The subtitle delay appears with every player I have tried: Jellyfin, VLC and Kodi. There might be some other players as well, but all of them doesn't support included subtitles but would like to have them external.

**Comment by IlmariKu** [comment]
Yeah, it's strange, I'm experiencing the exact same behavior of 2 second delay with subtitles.

I need to personally get this fixed, so I also don't have to open up programs to manually adjust subtitles. I'll see if I find something and let the thread know.

**Comment by j57m-yh** [comment]
Same problem here. For me it seems that all downloads have subs 2 seconds early. Of course they can be delayed in player options, but as you need to do that every time starting to watch one of these downloads, it's a bit annoying.

In ffmpeg command line options list there is (quote):
"-itsoffset offset (input)    Set the input time offset.
offset must be a time duration specification, see (ffmpeg-utils) the Time duration section in the ffmpeg-utils(1) manual.
The offset is added to the timestamps of the input files. Specifying a positive offset means that the corresponding streams are delayed by the time duration specified in offset."

Could this be a way to modify the timing of subtitle stream(s) if I knew (which I obviously do not) how to feed to ffmpeg?

**Comment by andeem** [comment]
Subtitle timing seems to be still off.

I run yle-dl with -VV flag and did notice something that might be relevant.
In the output, there's this ffprobe(?) line:
`  Duration: 00:29:10.04, **start: 2.000000**, bitrate: 0 kb/s`
The start value seems to correlate with the subtitle offset. I tested with few episodes/movies and those that were published to Areena quite recently (earliest was at the start of August 2024) did have the 2 second offset.

But the this one https://areena.yle.fi/1-64122776, published at February 2024, had different offset.
`  Duration: 00:44:55.14, **start: 0.101000**, bitrate: 0 kb/s`
Subtitles seem to be slightly off, but 0.1s is so small offset that it's hard to tell.

**Comment by j57m-yh** [comment]
Some observations about subtitle timing vs. publish date:
2023-04-17  subs ok
2023-07-31  subs early
2024-01-04  subs ok
2024-02-08  subs ok
2024-09-09  subs early.
What the actual programs those were - sorry, didn't write those down, can't remember anymore...
Anyway, it seems the publishing date cannot be trusted in judging, whether the timing is OK or not.

The ffmpeg -itsoffset command line option added to the actual download command sent to ffmpeg corrected the timing at download, I tested that. But then you would need to know in advance if there is a timing "error" in the stream - and because publishing date gives no reliable clue of that, you'll have to download first to be able to check. So, not very useful.

I noticed that "start: n.nnn" in the -VV output also, but for some reason I assumed it's the time the first sub line appears counted from the start of the video and didn't check it any further. So this is still only guessing.

**Comment by user-1-2-3-4-5** [comment]
https://areena.yle.fi/1-50318088

The subtitles appear around 2,5 seconds early, but I haven't checked other titles recently.

The Fedora-packaged version `20250730-3.fc43`, and the same result with the Python package (via `pipx`).

**Comment by j57m-yh** [comment]
Since late 2024 I haven't come across a single title with correct timing. The ffmpeg command line option "-itsoffset" did repair the timing error at download when I tested it, but that means you have write the download command line manually which is too complicated to be a practical solution.
(yle-dl = 20250730)

**Comment by tmu-k** [comment]
Yeah I have been struggling with subtitles also for at least a year... And "fixing" them manually with mkvtoolnix. For me the delay is always exactly 2 seconds (with movies & TV episodes). I created PR to add --subdelay argument which should at least remove the need to adjust subtitle timings in separate tools...

**Comment by j57m-yh** [comment]
Looks like the option I've been dreaming about for over a year...

**Comment by aajanki** [comment]
Merged @tmu-k 's PR. Thanks!

**Comment by aajanki** [comment]
The latest release 20260520 automatically adjusts subtitle timing. If the automatic detection fails, the delay can be manually specified with the `--subdelay` argument.
